File: python/sodl_weights/weight_manifest.py
"""Weight-manifest lifecycle helpers backed by SODL and optional native acceleration."""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterable, Mapping

from . import _rust_bridge
from sodl_weights.store import BlobStore


@dataclass(slots=True)
class WeightManifestCluster:
    cluster_id: int
    blob_id: str
    member_token_ids: list[int]


@dataclass(slots=True)
class WeightManifest:
    origin_id: str
    vocab_size: int
    embedding_dim: int
    clusters: list[WeightManifestCluster]
    metadata: dict[str, Any]


def _cluster_from_mapping(payload: Mapping[str, Any]) -> WeightManifestCluster:
    return WeightManifestCluster(
        cluster_id=int(payload["cluster_id"]),
        blob_id=str(payload["blob_id"]),
        member_token_ids=[int(item) for item in (payload.get("member_token_ids") or [])],
    )


def _manifest_from_mapping(payload: Mapping[str, Any]) -> WeightManifest:
    return WeightManifest(
        origin_id=str(payload.get("origin_id") or ""),
        vocab_size=int(payload.get("vocab_size", 0)),
        embedding_dim=int(payload.get("embedding_dim", 0)),
        clusters=[_cluster_from_mapping(item) for item in (payload.get("clusters") or [])],
        metadata=dict(payload.get("metadata") or {}),
    )


class _PythonWeightManifestBackend:
    def __init__(self, manifest_path: str | Path, blob_root: str | Path | None = None) -> None:
        self._manifest_path = Path(manifest_path)
        self._manifest_path.parent.mkdir(parents=True, exist_ok=True)
        self._blob_store = BlobStore(blob_root) if blob_root is not None else None

    def load_manifest(self) -> WeightManifest | None:
        if not self._manifest_path.exists():
            return None
        payload = json.loads(self._manifest_path.read_text(encoding="utf-8"))
        if not isinstance(payload, dict):
            raise ValueError(f"Weight manifest must be a JSON object: {self._manifest_path}")
        return _manifest_from_mapping(payload)

    def resolve_origin_id(
        self,
        checkpoint_origin: str,
        resume_record: Mapping[str, Any] | None = None,
    ) -> tuple[str | None, str]:
        metadata = dict((resume_record or {}).get("metadata") or {})
        resume_origin_id = str(metadata.get("sodl_origin_id") or "").strip()
        if resume_origin_id:
            return resume_origin_id, "resume_record"

        manifest = self.load_manifest()
        if manifest is None:
            return None, "new"
        origin_id = str(manifest.origin_id).strip()
        if not origin_id:
            return None, "new"
        recorded_checkpoint_origin = str(manifest.metadata.get("checkpoint_origin") or "").strip()
        if recorded_checkpoint_origin and recorded_checkpoint_origin != checkpoint_origin:
            return None, "new"
        return origin_id, "manifest"

    def write_manifest(
        self,
        origin_id: str,
        vocab_size: int,
        embedding_dim: int,
        clusters: list[WeightManifestCluster],
        metadata: Mapping[str, Any] | None = None,
    ) -> WeightManifest:
        previous_manifest = self.load_manifest()
        resolved_metadata = dict((previous_manifest.metadata if previous_manifest is not None else {}) or {})
        if metadata is not None:
            resolved_metadata.update(dict(metadata))
        manifest = WeightManifest(
            origin_id=str(origin_id),
            vocab_size=int(vocab_size),
            embedding_dim=int(embedding_dim),
            clusters=sorted(clusters, key=lambda item: int(item.cluster_id)),
            metadata=resolved_metadata,
        )
        self._manifest_path.write_text(
            json.dumps(
                {
                    "origin_id": manifest.origin_id,
                    "vocab_size": manifest.vocab_size,
                    "embedding_dim": manifest.embedding_dim,
                    "clusters": [
                        {
                            "cluster_id": cluster.cluster_id,
                            "blob_id": cluster.blob_id,
                            "member_token_ids": list(cluster.member_token_ids),
                        }
                        for cluster in manifest.clusters
                    ],
                    "metadata": manifest.metadata,
                },
                indent=2,
            ),
            encoding="utf-8",
        )
        if previous_manifest is not None and self._blob_store is not None:
            live_blob_ids = {cluster.blob_id for cluster in manifest.clusters}
            stale_blob_ids = {
                cluster.blob_id
                for cluster in previous_manifest.clusters
                if cluster.blob_id not in live_blob_ids
            }
            for blob_id in sorted(stale_blob_ids):
                try:
                    self._blob_store.delete(blob_id)
                except FileNotFoundError:
                    continue
        return manifest


class WeightManifestStore:
    def __init__(self, manifest_path: str | Path, blob_root: str | Path | None = None) -> None:
        native = _rust_bridge.create_weight_manifest_store(
            str(manifest_path),
            str(blob_root) if blob_root is not None else None,
        )
        self._native = native
        self._fallback = None if native is not None else _PythonWeightManifestBackend(
            manifest_path,
            blob_root,
        )

    def _backend(self) -> Any:
        return self._native if self._native is not None else self._fallback

    def load_manifest(self) -> WeightManifest | None:
        backend = self._backend()
        if self._native is not None:
            payload = backend.load_manifest()
            return None if payload is None else _manifest_from_mapping(json.loads(payload))
        return backend.load_manifest()

    def resolve_origin_id(
        self,
        checkpoint_origin: str,
        resume_record: Mapping[str, Any] | None = None,
    ) -> tuple[str | None, str]:
        backend = self._backend()
        if self._native is not None:
            payload = json.loads(
                backend.resolve_origin_id(
                    checkpoint_origin,
                    json.dumps(dict(resume_record)) if resume_record is not None else None,
                )
            )
            origin_id = payload.get("origin_id")
            return (None if origin_id in (None, "") else str(origin_id), str(payload.get("source") or "new"))
        return backend.resolve_origin_id(checkpoint_origin, resume_record)

    def write_manifest(
        self,
        origin_id: str,
        vocab_size: int,
        embedding_dim: int,
        clusters: Iterable[WeightManifestCluster],
        metadata: Mapping[str, Any] | None = None,
    ) -> WeightManifest:
        cluster_list = list(clusters)
        backend = self._backend()
        if self._native is not None:
            payload = json.loads(
                backend.write_manifest(
                    origin_id,
                    int(vocab_size),
                    int(embedding_dim),
                    json.dumps(
                        [
                            {
                                "cluster_id": int(cluster.cluster_id),
                                "blob_id": str(cluster.blob_id),
                                "member_token_ids": [int(item) for item in cluster.member_token_ids],
                            }
                            for cluster in cluster_list
                        ]
                    ),
                    json.dumps(dict(metadata or {})),
                )
            )
            return _manifest_from_mapping(payload)
        return backend.write_manifest(origin_id, vocab_size, embedding_dim, cluster_list, metadata)


def export_manifest_clusters(
    token_index: Any,
    cluster_blob_ids: Mapping[int, str],
) -> list[WeightManifestCluster]:
    return [
        WeightManifestCluster(
            cluster_id=int(cluster_id),
            blob_id=str(blob_id),
            member_token_ids=[int(token_id) for token_id in token_index.cluster_members(int(cluster_id))],
        )
        for cluster_id, blob_id in sorted(cluster_blob_ids.items(), key=lambda item: int(item[0]))
    ]

File: crates/sodl-service/src/weight_manifest.rs
use std::collections::HashSet;
use std::fs;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use serde_json::Value;
use sodl_cas::{BlobStore, FsBlobStore};
use sodl_core::{BlobId, Result, SodlError};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WeightManifestCluster {
    pub cluster_id: u64,
    pub blob_id: BlobId,
    pub member_token_ids: Vec<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WeightManifest {
    pub origin_id: String,
    #[serde(default)]
    pub vocab_size: usize,
    #[serde(default)]
    pub embedding_dim: usize,
    #[serde(default)]
    pub clusters: Vec<WeightManifestCluster>,
    #[serde(default = "default_json_object")]
    pub metadata: Value,
}

pub struct WeightManifestStore {
    manifest_path: PathBuf,
    blob_store: Option<FsBlobStore>,
}

impl WeightManifestStore {
    pub fn open(
        manifest_path: impl AsRef<Path>,
        blob_root: Option<impl AsRef<Path>>,
    ) -> Result<Self> {
        let manifest_path = manifest_path.as_ref().to_path_buf();
        if let Some(parent) = manifest_path.parent() {
            fs::create_dir_all(parent)
                .map_err(|err| SodlError::Io(format!("create weight manifest dir: {err}")))?;
        }
        let blob_store = blob_root
            .map(|root| FsBlobStore::open(root.as_ref()))
            .transpose()?;
        Ok(Self {
            manifest_path,
            blob_store,
        })
    }

    pub fn load_manifest(&self) -> Result<Option<WeightManifest>> {
        if !self.manifest_path.exists() {
            return Ok(None);
        }
        let bytes = fs::read(&self.manifest_path).map_err(|err| {
            SodlError::Io(format!(
                "read weight manifest {}: {err}",
                self.manifest_path.display()
            ))
        })?;
        let manifest: WeightManifest = serde_json::from_slice(&bytes).map_err(|err| {
            SodlError::Serialization(format!(
                "parse weight manifest {}: {err}",
                self.manifest_path.display()
            ))
        })?;
        Ok(Some(manifest))
    }

    pub fn resolve_origin_id(
        &self,
        checkpoint_origin: &str,
        resume_record: Option<&Value>,
    ) -> Result<(Option<String>, String)> {
        let metadata = resume_record
            .and_then(|payload| payload.get("metadata"))
            .and_then(Value::as_object);
        if let Some(origin_id) = metadata
            .and_then(|meta| meta.get("sodl_origin_id"))
            .and_then(Value::as_str)
            .map(str::trim)
            .filter(|value| !value.is_empty())
        {
            return Ok((Some(origin_id.to_string()), "resume_record".to_string()));
        }

        let Some(manifest) = self.load_manifest()? else {
            return Ok((None, "new".to_string()));
        };
        let origin_id = manifest.origin_id.trim();
        if origin_id.is_empty() {
            return Ok((None, "new".to_string()));
        }
        let recorded_checkpoint_origin = manifest
            .metadata
            .get("checkpoint_origin")
            .and_then(Value::as_str)
            .map(str::trim)
            .unwrap_or("");
        if !recorded_checkpoint_origin.is_empty() && recorded_checkpoint_origin != checkpoint_origin
        {
            return Ok((None, "new".to_string()));
        }
        Ok((Some(origin_id.to_string()), "manifest".to_string()))
    }

    pub fn write_manifest(
        &self,
        origin_id: &str,
        vocab_size: usize,
        embedding_dim: usize,
        clusters: Vec<WeightManifestCluster>,
        metadata: Option<Value>,
    ) -> Result<WeightManifest> {
        let previous_manifest = self.load_manifest()?;
        let mut resolved_metadata = previous_manifest
            .as_ref()
            .map(|manifest| manifest.metadata.clone())
            .unwrap_or_else(default_json_object);
        if let Some(next_metadata) = metadata {
            merge_metadata(&mut resolved_metadata, next_metadata);
        }

        let mut clusters = clusters;
        clusters.sort_by_key(|item| item.cluster_id);
        let manifest = WeightManifest {
            origin_id: origin_id.to_string(),
            vocab_size,
            embedding_dim,
            clusters,
            metadata: resolved_metadata,
        };

        let tmp_path = self.manifest_path.with_extension("json.tmp");
        let bytes = serde_json::to_vec_pretty(&manifest)
            .map_err(|err| SodlError::Serialization(format!("serialize weight manifest: {err}")))?;
        fs::write(&tmp_path, &bytes).map_err(|err| {
            SodlError::Io(format!(
                "write weight manifest temp {}: {err}",
                tmp_path.display()
            ))
        })?;
        fs::rename(&tmp_path, &self.manifest_path).map_err(|err| {
            SodlError::Io(format!(
                "replace weight manifest {}: {err}",
                self.manifest_path.display()
            ))
        })?;

        if let (Some(previous), Some(blob_store)) = (previous_manifest, &self.blob_store) {
            let live_blob_ids: HashSet<BlobId> = manifest
                .clusters
                .iter()
                .map(|cluster| cluster.blob_id.clone())
                .collect();
            let stale_blob_ids: HashSet<BlobId> = previous
                .clusters
                .into_iter()
                .map(|cluster| cluster.blob_id)
                .filter(|blob_id| !live_blob_ids.contains(blob_id))
                .collect();
            for blob_id in stale_blob_ids {
                if blob_store.has(&blob_id)? {
                    blob_store.delete(&blob_id)?;
                }
            }
        }

        Ok(manifest)
    }
}

fn default_json_object() -> Value {
    Value::Object(serde_json::Map::new())
}

fn merge_metadata(target: &mut Value, incoming: Value) {
    let Value::Object(target_map) = target else {
        *target = default_json_object();
        merge_metadata(target, incoming);
        return;
    };
    if let Value::Object(incoming_map) = incoming {
        for (key, value) in incoming_map {
            target_map.insert(key, value);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn manifest_roundtrip_and_gc() {
        let tmp = tempfile::tempdir().unwrap();
        let manifest_path = tmp.path().join("run").join("sodl_manifest.json");
        let blob_root = tmp.path().join("blobs");
        let store = WeightManifestStore::open(&manifest_path, Some(&blob_root)).unwrap();
        let blob_store = FsBlobStore::open(&blob_root).unwrap();

        let blob_a = BlobId(
            "blake3:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa".to_string(),
        );
        let blob_b = BlobId(
            "blake3:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb".to_string(),
        );
        let blob_c = BlobId(
            "blake3:cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc".to_string(),
        );
        blob_store
            .put(&blob_a, bytes::Bytes::from_static(b"a"))
            .unwrap();
        blob_store
            .put(&blob_b, bytes::Bytes::from_static(b"b"))
            .unwrap();
        blob_store
            .put(&blob_c, bytes::Bytes::from_static(b"c"))
            .unwrap();

        store
            .write_manifest(
                "origin:1",
                2,
                4,
                vec![
                    WeightManifestCluster {
                        cluster_id: 0,
                        blob_id: blob_a.clone(),
                        member_token_ids: vec![0],
                    },
                    WeightManifestCluster {
                        cluster_id: 1,
                        blob_id: blob_b.clone(),
                        member_token_ids: vec![1],
                    },
                ],
                Some(serde_json::json!({"checkpoint_origin":"carla-large","note":"keep"})),
            )
            .unwrap();
        let updated = store
            .write_manifest(
                "origin:1",
                2,
                4,
                vec![
                    WeightManifestCluster {
                        cluster_id: 0,
                        blob_id: blob_a.clone(),
                        member_token_ids: vec![0],
                    },
                    WeightManifestCluster {
                        cluster_id: 1,
                        blob_id: blob_c.clone(),
                        member_token_ids: vec![1],
                    },
                ],
                None,
            )
            .unwrap();

        assert_eq!(updated.metadata["note"], "keep");
        assert!(blob_store.has(&blob_a).unwrap());
        assert!(!blob_store.has(&blob_b).unwrap());
        assert!(blob_store.has(&blob_c).unwrap());
    }

    #[test]
    fn resolve_origin_prefers_resume_record_then_manifest() {
        let tmp = tempfile::tempdir().unwrap();
        let manifest_path = tmp.path().join("sodl_manifest.json");
        let store = WeightManifestStore::open(&manifest_path, None::<&Path>).unwrap();
        store
            .write_manifest(
                "origin:manifest",
                1,
                1,
                vec![],
                Some(serde_json::json!({"checkpoint_origin":"carla-large"})),
            )
            .unwrap();

        let resolved = store
            .resolve_origin_id(
                "carla-large",
                Some(&serde_json::json!({"metadata":{"sodl_origin_id":"origin:resume"}})),
            )
            .unwrap();
        assert_eq!(resolved.0.as_deref(), Some("origin:resume"));
        assert_eq!(resolved.1, "resume_record");

        let resolved = store.resolve_origin_id("carla-large", None).unwrap();
        assert_eq!(resolved.0.as_deref(), Some("origin:manifest"));
        assert_eq!(resolved.1, "manifest");

        let resolved = store.resolve_origin_id("carla-medium", None).unwrap();
        assert_eq!(resolved.0, None);
        assert_eq!(resolved.1, "new");
    }
}
